datascience/maps.py

```diff
@@ -0,0 +1,346 @@
+                    'border: none"></iframe>' % (map_html, map.width, map.height))
+    
+    def before_to_html(self):
+        """
+        Called before to_html, add points and regions before outputting map 
+        """
+        for point in self.data['points']:
+            point.map(self.folium)
+
+        for region in self.data['regions']:
+            region.map(self.folium)
+        
+
+class MapPoint(MapEntity):
+    """ A circle. Draw by passing into draw_map."""
+    
+    mapper = 'circle_marker'
+    
+    defaults = {
+        'location': [],             # location -- lat-long pair at center of circle
+        'radius': 10,               # radius -- radius of circle on map
+        'popup': '',                # popup -- text that pops up when circle is clicked
+        'line_color': '#3186cc',    # line_color -- color of circle border
+        'fill_color': '#3186cc',    # fill_color -- color of circle within border
+        'fill_opacity': 0.6         # fill_opacity -- opacity of circle fill
+    }
+
+    def __init__(self, location, **kwargs):
+        super().__init__(location=location, **kwargs)
+
+    def __str__(self):
+        return str(list(self.location))
+
+
+class MapRegion(MapEntity):
+    """
+    A polygon. Draw by passing into draw_map.
+
+    regions -- list of MapRegions (if present, locations and points are not used)
+    locations -- list of list of lat-long pairs for locations with multiple 
+                 polygons (if present, points are not used)
+    points -- list of lat-long pairs for a normal polygon
+    """
+    # composite options
+
+    CHILDREN = 'children'  # invoke map() for children
+    GROUP = 'group'  # invoke to_json() for children
+    ONE = 'one'  # invoke to_polygon() for children
+    
+    mapper = 'geo_json'
+
+    dataonly = ['points', 'locations', 'regions', 'composite']
+    
+    defaults = {
+        'composite': GROUP,
+        'regions': [],
+        'locations': [[]],
+        'points': [],
+        'fill_color': 'blue',
+        'fill_opacity': 0.6,
+        'line_color': 'black',
+        'line_weight': 1,
+        'line_opacity': 1
+    }
+
+    TEMP_FILE = None  # TODO: temporary workaround - will need a more permanent solution
+
+    allowed_collections = (list, set, tuple)
+
+    def before_map(self):
+        """ Prepares for mapping by passing JSON representation """
+        self._validate()
+        json_str = json.dumps(self._to_json()).replace(']]]}}', ']]]}}\n')
+        # self.attributes['geo_str'] = json_str
+        if self['locations'][0] or self['points'] or self['regions']:  # TODO: temporary workaround
+            import os
+            try:
+                os.mkdir('cache/')
+            except FileExistsError:
+                pass
+            self.TEMP_FILE = 'cache/'+str(round(random.random()*100))+'temp.json'
+            open(self.TEMP_FILE, 'w').write(json_str)
+            self.attributes['geo_path'] = self.TEMP_FILE
```

Comment:
@SamLau95 I couldn't get `geo_str` to work; do you know what it's supposed to take?

I know the `geoJSON` is valid, so I employed an ugly hack... for now - wrote it to a file and had folium read from the file.